rfd/0194-k8s-rbac-simplification.md

```diff
@@ -0,0 +1,220 @@
+---
+authors: Guillaume Charmes (guillaume.charmes@goteleport.com)
+state: draft
+---
+
+# RFD 0194 - Kubernetes RBAC Simplification
+
+## Required Approvers
+
+- Engineering: @rosstimothy && @hugoShaka && @tigrato
+
+## What
+
+The purpose of this RFD is to propose a simplification of the Kubernetes
+RBAC (Role-Based Access Control) within the Teleport project.
+
+The goal is to streamline the user experience and reduce the complexity of
+getting up and running on day 1 and managing permissions later on.
+
+## Why
+
+Today, the Kubernetes RBAC in Teleport is complex and can be difficult for
+users to manage.
+A proper setup requires some external Kubernetes config and matching role
+setup.
+This complexity leads to misconfigurations and security issues.
+
+Simplifying the RBAC model will help improve usability and security.
+
+### References
+
+A common issue currently is the un-intuitive result of complex rule sets and
+more critically, difficult to get setup on day 1.
+
+- @klizhentas struggling to setup a cluster following the `self-hosted` flow:
+  (internal) <https://gravitational.slack.com/archives/C03SULRAAG3/p1715394587181739>
+- Customer confusion on role resolution:
+  - <https://github.com/gravitational/teleport/issues/45475>
+  - (internal) <https://goteleport.slack.com/archives/C0582MBSMHN/p1738187497666819>
+
+## Goals
+
+- Make the Teleport agent authoritative
+  - Handle permissions for the users instead of relying on impersonating a
+    different user/group
+- Discourage (or even deprecate?) the use of the `kubernetes_groups` and
+  `kubernetes_users` in favor of `kubernetes_resources`
+- Clarify / Improve documentation for the various ways to setup Kubernetes in
+  Teleport.
+
+## Proposal
+
+### Background
+
+Currently, when enrolling a Kubernetes cluster and setting it up to use the
+role `system:masters`, everything works as expected. Teleport is super admin
+and yields reduced permissions to the user based on the `kubernetes_resources`
+and `kubernetes_label` fields from the Teleport role.
+
+### Impersonation
+
+To limit the number of code changes, the proposal is to default the
+`kubernetes_group` value to `system:masters` and hide it from the role.
+This can be done by allowing the field to be `null`, and treating `null` as
+`systems:masters`.
+
+The `kubernetes_user` should default to the Teleport user auth email.
+
+### Labels
+
+We currently support labels at the cluster level, however, resources only
+support name/namespace filtering.
+Adding an optional field for labels would allow for more flexible and versatile
+control over permissions, allowing something like:
+
+```yaml
+kubernetes_resources:
+  - kind: '*'
+    name: '*'
+    namespace: '*'
+    verbs: ['*']
```

Comment:
I would expand the resources with the following structure. Adding subresources makes sense to prevent users from exec, logs, cp...

Splitting the API and kind also improves readability 

```suggestion
kubernetes_resources:
  - kind: '*'
    api: 'API'
    name: '*'
    namespace: '*'
    verbs: ['*']
    sub_resources: ['*']
```

Reply:
+1 on exposing the api field.

I don't like the sub-resource thing because people keep accidentally allowing pod exec in their kube read-only roles when they grant get/list/watch access on the kind '*'. This happened twice this year and even after explaining how kube RBAC is implemented to the customers, I don't think they really got what their mistake was and I'm pretty sure they will do it again.

If we do this, I think that we should at least keep a special case for pod execution and port forwarding as the virtual verbs we added are a saner UX than what kube offers.

Reply:
The subresource will ignore the verb in some cases so people won't shoot themselves